sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,358 @@
+        if (date.isEmpty) {
+          logInfo("[CBO] Date literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+        date.get.toString
+      case t: TimestampType if literal.dataType.isInstanceOf[StringType] =>
+        val timestamp = DateTimeUtils.stringToTimestamp(literal.value.asInstanceOf[UTF8String])
+        if (timestamp.isEmpty) {
+          logInfo("[CBO] Timestamp literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+        timestamp.get.toString
+      case _ => literal.value.toString
+    }
+
+    val minString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+    val maxString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.max match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+
+    /**
+     * decide if the value is in [min, max] of the column.
+     * We currently don't store min/max for binary/string type.
+     * Hence, we assume it is in boundary for binary/string type.
+     */
+    val inBoundary: Boolean = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) =>
+            datumString.toDouble >= minString.toDouble && datumString.toDouble <= maxString.toDouble
+          case None => true
+        }
+      case _ => true
+    }
+
+    if (inBoundary) {
+      percent = 1.0 / ndv.toDouble
+
+      if (update) {
+        /**
+         * We update ColumnStat structure after apply this equality predicate.
+         * Set distinctCount to 1.  Set nullCount to 0.
+         */
+        val oneBigInt: BigInt = 1
+        val zeroBigInt: BigInt = 0
+        val newStats = attrRef.dataType match {
+          case dataType: DataType if (dataType.isInstanceOf[NumericType]
+            || dataType.isInstanceOf[DateType]
+            || dataType.isInstanceOf[TimestampType]) =>
+            val newValue = Some(datumString.toDouble)
+            aColStat.copy(distinctCount = oneBigInt, min = newValue,
+              max = newValue, nullCount = zeroBigInt)
+          case _ => aColStat.copy(distinctCount = oneBigInt, nullCount = zeroBigInt)
+        }
+        mutableColStats += (attrRef.name -> newStats)
+      }
+
+    } else {
+      percent = 0.0
+    }
+
+    percent
+  }
+
+  def evaluateInSet(

```

Comment:
With only basic column stats, how to get rowCount after such filter? I suppose you will use #valuesInRange / ndv, this could be much smaller than the actual result, so it's risky.